### ISSUE TITLE
Roslyn doesn't respect flow annotation when checking ref arguments' nullability mismatch

### ISSUE THREAD

**Issue by TessenR** [dotnet/roslyn#42743]
**Version Used**: 

```
Branch master (10 Mar 2020)
Latest commit 441c154 by msftbot[bot]:
Merge pull request #42303 from CyrusNajmabadi/useCompoundParens

Support parenthesized expressions in use-compound-operator.
```

**Steps to Reproduce**:

Compile and run the following code
```cs
#nullable enable
using System.Diagnostics.CodeAnalysis;

class C
{
  public static void Main()
  {
    M(out string s);
    s.ToString();
  }
    
  static void M([NotNull] out string? s)
  {
    s = "";
    Ref(ref s);
    
    void Ref(ref string? s) => s = null;
  }
}
```

[sharplab.io](https://sharplab.io/#v2:EYLgtghgzgLgpgJwDQxAgrgOyQExAagB8BiTdAG3ImHLgAI5NraBYAKAAEAGOjgRgB0AEQCWEAOaYA9rBEBjKAIDCUnHACCTcgE8oIqAG527DgCY6S9gG92dXqdO26NtnbsBZABRT0MXnx4oAEojVzdFABUpAGUYBBFMcU8QpwBfJzsnfgA2XgAWOi8AbQA5KRgSinIAXTofP34uAH46YKcXN1a6AF46ACI+0M6AJTgAM08EcdaUsMy5/LpRiamx/2aZnoA+Lt6ySiG6dLZUoA==)

**Expected Behavior**:
Warning for `ref s` argument. It might assign `null` to `s` which is marked as `[NotNull]`

**Actual Behavior**:
No warnings. The program crashes at runtime with a `NullReferenceException`

**Notes**
It seems that flow annotations are just ignored in this case, e.g. the following code has a false warning:
```cs
#nullable enable
using System.Diagnostics.CodeAnalysis;

class C
{
  void M([MaybeNull] out string s)
  {
    s = null; // OK
    Out(out s); // OK
    Ref(ref s); // incorrect warning
    
    void Out(out string? s) => s = null;
    void Ref(ref string? s) => s = null;
  }
}
```
